# UPHL-BioNGS_Cecret_V3
Source Git Repository: https://github.com/UPHL-BioNGS/Cecret

By: [@erinyoung](https://github.com/erinyoung)

In order for this pipeline to run successfully you must provide the following files which are labeled with there association fields in the XML:

You need to provide samples to one of the following inputs; ```READS, SINGLE_READS, FASTAS, MULTIFASTAS``` Cecret can run with one or mix of samples of those types.

## READS
Paired end fastq files in a directory

## SINGLE_READS
Single end fastq files in a directory

## FASTAS
Fasta files in a directory

## MULTIFASTAS
Multifastas in a directory

## PROJECT_DIRS FROM EDITED GIT REPO
The directories found in this Github directory:
```
bin
configs
data
modules
subworkflows
```


